huaweicloud/resource_huaweicloud_networking_secgroup_rule.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/fmtp"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/logp"
+)
+
+func ResourceNetworkingSecGroupRuleV3() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceNetworkingSecGroupRuleV3Create,
+		ReadContext:   resourceNetworkingSecGroupRuleV3Read,
+		DeleteContext: resourceNetworkingSecGroupRuleV3Delete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"direction": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"ingress", "egress",
+				}, true),
+			},
+			"ethertype": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"IPv4", "IPv6",
+				}, true),
+			},
+			"security_group_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"ports": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+			},
+			"protocol": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+				ValidateFunc: validation.Any(
+					validation.StringInSlice([]string{"tcp", "udp", "icmp", "icmpv6"}, false),
+					validation.StringMatch(regexp.MustCompile("^([0-1]?[0-9]?[0-9]|2[0-4][0-9]|25[0-5])$"),
+						"The valid protocol is range from 0 to 255.",
+					),
+				),
+			},
+			"remote_group_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+			},
+			"remote_ip_prefix": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+				StateFunc: func(v interface{}) string {
+					return strings.ToLower(v.(string))
+				},
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceNetworkingSecGroupRuleV3Create(ctx context.Context, d *schema.ResourceData,
+	meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.NetworkingV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating HuaweiCloud networking v3 client: %s", err)
+	}
+
+	protocol := d.Get("protocol").(string)
+	if _, ok := d.GetOk("ports"); ok && protocol == "" {
+		return fmtp.DiagErrorf("A protocol must be specified when using port_range_min and port_range_max")
```

Comment:
the message is misleading.
also you can use `Required` in schema to check it.

Reply:
done.
1. The **port_range_min** and **port_range_max** needs to be set.
2. The **ports** is conflict with **port_range_min** and **port_range_max**.
3. The **protocol** is required if **ports** or **port_range_min** or **port_range_max** is set.